### ISSUE TITLE
Image Copy looses layer annotations when reusing blob

### ISSUE THREAD

**Issue by mikenorgate** [containers/image#1888]
I've noticed an issue when pushing an image to a container registry where the layer annotations are lost if the layer already exists in the registry

As an example a layer with this digest
```
{
    "mediaType": "application/vnd.oci.image.layer.v1.tar+zstd",
    "digest": "sha256:14e05236c928e6b9583038025d14979e596210b54c4a2966bfff41172e7acab4",
    "size": 1199,
    "annotations": {
        "io.github.containers.zstd-chunked.manifest-checksum": "sha256:6a716ddd20d5e0a811ed8530a6fd5ed10eac5fd06a80d4690a0b973857a940b4",
        "io.github.containers.zstd-chunked.manifest-position": "748:403:1976:1"
    }
}
```
results in
```
{
    "mediaType": "application/vnd.oci.image.layer.v1.tar+zstd",
    "digest": "sha256:14e05236c928e6b9583038025d14979e596210b54c4a2966bfff41172e7acab4",
    "size": 1199
}
```

After digging through the code I believe it's down to this function not knowing about the original annotations https://github.com/containers/image/blob/0c6d07e02a9a0c8119885cc461c89b7452684ba5/docker/docker_image_dest.go#L321

**Comment by mtrmac** [comment]
Thanks for your report.

Yes, that’s clearly a bug.

The right location for a fix is a bit less clear. The documented semantics of `PutBlob` is that the return value should return “at least” `Digest` and `Size, which I think makes sense — that way only the generic code needs to care about annotations, not the individual transports.

But we should also clearly document what happens with the other fields of `types.BlobInfo`.

**Comment by mtrmac** [comment]
#1892 should fix this; independent testing would be appreciated.

**Comment by mikenorgate** [comment]
@mtrmac I have just tested your changes and I'm seeing the same behaviour

For a bit of debugging in seems like the annotation information is being lost on this line https://github.com/containers/image/blob/f9ef62d154120aa7ba41026c89aa7a1237204961/copy/single.go#L371

The value of `srcInfos` at this line https://github.com/containers/image/blob/f9ef62d154120aa7ba41026c89aa7a1237204961/copy/single.go#L360 is
![image](https://user-images.githubusercontent.com/2479870/227161591-3e33d015-324f-45fd-82b7-d016a24f5449.png)

The value of `updatedSrcInfos` at this line https://github.com/containers/image/blob/f9ef62d154120aa7ba41026c89aa7a1237204961/copy/single.go#L362 is
![image](https://user-images.githubusercontent.com/2479870/227161795-0717fcf1-71b5-4450-b127-9be7f2c52e02.png)

**Comment by mtrmac** [comment]
Thanks. To confirm, is that a push from `containers-storage:`?

**Comment by mtrmac** [comment]
Oh — I have just noticed these are the Zstd annotations. Those are specific to that _exact byte representation_; i.e. they shouldn’t be preserved when we replace one compressed version (one with Zstd:chunked data) with another (one with Zstd:chunked data in other locations, or with no Zstd:chunked data at all).

The unspecified semantics of the `annotations` field are going to be a real hassle…

---

And in the specific case of a push from `containers-storage:` (if that’s what this is), the compressed version is never recorded in c/storage, and preserving the Zstd:chunked annotations doesn’t make any sense; the :chunked data must be recreated during the push, from scratch — and that can add (some new version of) the Zstd annotations.

**Comment by mtrmac** [comment]
A truly correct fix would be quite involved; for now, #1894 just adds some FIXMEs, and a long commit message, to have a permanent record.
